### PR TITLE
YAML test framework: separate `skip` and `requires` sections 

### PR DESCRIPTION
**Pull request by ldematte**
This PR introduces a new `requires` field in parallel to the existing `skip` field; after trying out more elaborate alternatives, I choose to keep the change minimal wrt https://github.com/elastic/elasticsearch/pull/103909 and previous refactoring. Since renaming `SkipSection` to `PrerequisiteSection` (as it was strange to keep the name "Skip" when it contains both `skip` and `requires`) is making things muddier, this is what changed:
- parseIfNext/parseInternal is now "detecting" both `skip` and `requires` (in any order)
- parseInternal is split into (acutally, delegates) parseSkipSection and parseRequiresSection
- parseSkipSection and parseRequiresSection are the same as before -- except each accepts and parses a sub-selection of fields.
- the builder introduced previously is the same, and it is shared by parseSkipSection and parseRequiresSection; this keeps the change minimal but also make cross-field validation easier.

In a first iteration, I think about keeping `skip: features: ...` so we do not have to mix a monster change with this PR; I have added `requires: test_runner_features: ...` in parallel. We can migrate tests to that separately.

See `PrerequisiteSectionTests` for how it looks like. An example: 

```
requires:
   test_runner_features:  some-runner feature
   cluster_features:      needed-feature
   reason:                test needs needed-feature to run
skip:
   cluster_features:      undesired-feature
   reason:                test cannot run when undesired-feature are present
```

### CODE REVIEW

File: test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/PrerequisiteSection.java [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+        List<String> yamlRunnerFeatures
+    ) {
+        this.skipCriteriaList = skipCriteriaList;
+        this.requireCriteriaList = requireCriteriaList;
+        this.yamlRunnerFeatures = yamlRunnerFeatures;
+        this.skipReason = skipReason;
+        this.requireReason = requireReason;
+    }
+
+    public boolean yamlRunnerHasFeature(String feature) {
```
Comment (mosche) [inline test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/PrerequisiteSection.java:359]: nit, maybe `yamlRunnerRequiresFeature` is easier to understand?

File: test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/ClientYamlTestSuite.java [hunk 2]
```diff
@@ -292,13 +292,13 @@ private static boolean hasSkipFeature(
...
-            || (setupSection != null && hasSkipFeature(feature, setupSection.getSkipSection()))
-            || (teardownSection != null && hasSkipFeature(feature, teardownSection.getSkipSection()));
+        return (testSection != null && hasSkipFeature(feature, testSection.getPrerequisiteSection()))
+            || (setupSection != null && hasSkipFeature(feature, setupSection.getPrerequisiteSection()))
+            || (teardownSection != null && hasSkipFeature(feature, teardownSection.getPrerequisiteSection()));
     }
 
-    private static boolean hasSkipFeature(String feature, SkipSection skipSection) {
-        return skipSection != null && skipSection.yamlRunnerHasFeature(feature);
+    private static boolean hasSkipFeature(String feature, PrerequisiteSection prerequisiteSection) {
```
Comment (mosche) [inline test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/ClientYamlTestSuite.java:300]: Should this be renamed? `hasYamlRunnerFeature` or similar

File: test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/PrerequisiteSection.java [hunk 3]
```diff
@@ -0,0 +1,401 @@
...
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+            if (token == XContentParser.Token.FIELD_NAME) {
+                currentFieldName = parser.currentName();
+            } else if (token.isValue()) {
+                if ("version".equals(currentFieldName)) {
+                    builder.skipIfVersion(parser.text());
+                } else if ("reason".equals(currentFieldName)) {
+                    builder.setSkipReason(parser.text());
+                } else if ("features".equals(currentFieldName)) {
+                    // TODO: legacy - remove
```
Comment (mosche) [inline test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/PrerequisiteSection.java:253]: Should we log a warning here and let people know how to migrate?

File: test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/PrerequisiteSection.java [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+                    builder.skipIfClusterFeature(parser.text());
+                } else {
+                    throw new ParsingException(
+                        parser.getTokenLocation(),
+                        "field " + currentFieldName + " not supported within skip section"
+                    );
+                }
+            } else if (token == XContentParser.Token.START_ARRAY) {
+                // TODO: legacy - remove
+                if ("features".equals(currentFieldName)) {
```
Comment (mosche) [inline test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/PrerequisiteSection.java:267]: Same as above, log a warning?

File: test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/PrerequisiteSection.java [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+    public boolean yamlRunnerHasFeature(String feature) {
+        return yamlRunnerFeatures.contains(feature);
+    }
+
+    boolean skipCriteriaMet(ClientYamlTestExecutionContext context) {
+        return skipCriteriaList.stream().anyMatch(c -> c.test(context));
+    }
+
+    boolean requiresCriteriaMet(ClientYamlTestExecutionContext context) {
+        return requireCriteriaList.stream().allMatch(c -> c.test(context));
```
Comment (mosche) [inline test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/PrerequisiteSection.java:368]: require`s`Criteria vs requireCriteria, not sure if that is intended

File: test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/PrerequisiteSection.java [hunk 6]
```diff
@@ -0,0 +1,401 @@
...
+        if (skipCriteriaMet(context)) {
+            throw new AssumptionViolatedException(buildMessage(testCandidateDescription, true));
+        }
+    }
+
+    boolean isEmpty() {
+        return skipCriteriaList.isEmpty() && requireCriteriaList.isEmpty() && yamlRunnerFeatures.isEmpty();
+    }
+
+    String buildMessage(String description, boolean isSkip) {
```
Comment (mosche) [inline test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/PrerequisiteSection.java:389]: 👍

File: test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/PrerequisiteSection.java [hunk 7]
```diff
@@ -0,0 +1,401 @@
...
+        StringBuilder messageBuilder = new StringBuilder();
+        messageBuilder.append("[").append(description).append("] skipped,");
+        var reason = isSkip ? skipReason : requireReason;
+        if (Strings.isNullOrEmpty(reason) == false) {
+            messageBuilder.append(" reason: [").append(reason).append("]");
+        }
+        if (yamlRunnerFeatures.isEmpty() == false) {
+            messageBuilder.append(" unsupported features ").append(yamlRunnerFeatures);
```
Comment (mosche) [inline test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/PrerequisiteSection.java:397]: Just a thought, how about appending that to the `requiresReason` in the builder when required runner features are missing?

**Review by mosche — COMMENTED** [review]
Thanks @ldematte, this is really looking good.
Wondering, what migration path do you have in mind?
Should we warn about the deprecation of `skip.features` and suggest what to use instead? 
Everything else is just some optional nits around naming, feel free to treat as you like.

File: test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/PrerequisiteSection.java (hunk 1 above)
Comment (ldematte) [reply test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/PrerequisiteSection.java:359]: For consistency with `ClientYamlTestSuite`, I have renamed all as `hasYamlRunnerFeature`

File: test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/PrerequisiteSection.java (hunk 3 above)
Comment (ldematte) [reply test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/PrerequisiteSection.java:253]: Added

File: test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/PrerequisiteSection.java (hunk 7 above)
Comment (ldematte) [reply test/yaml-rest-runner/src/main/java/org/elasticsearch/test/rest/yaml/section/PrerequisiteSection.java:409]: It's a good idea, but I prefer to leave it as is for now, so the output message does not change too much.

**Comment by mosche** [comment]
Thanks

**Review by mosche — APPROVED** [review]
LGTM 💯 
Sorry, for hitting the wrong button, fat fingers 🤦

**Comment by ldematte** [comment]
No problem :)
Re: the migration path, I think we'll do something similar to Version deprecation: our team will take some of it, and we'll advise other teams on how to migrate their part. The two can be done together: as we move from versions to features in YAML tests, we/other teams can take care of switching from `skip: features` to `requires: test_runner_features` too

**Pull request merged**